[JIT] Opinfo tests for nnc fusion

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71299
* __->__ #70465

These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

davidberard98 committed:
[WIP][JIT] Opinfo test for nnc fusion

These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8520,6 +8523,7 @@ def ref_pairwise_distance(input1, input2):
            assert_autodiffed=True,
            supports_forward_ad=True,
            supports_fwgrad_bwgrad=True,
+           expect_nnc_fusion=False,
```
@eellison lmk if you have any general comments on this. If you think it looks good I'll go through the list of ops and disable the ones where nnc fusion currently fails with `expect_nnc_fusion=False` like here.

Krovatkin commented on test/test_ops.py:
```diff
@@ -1155,6 +1156,68 @@ def _fn(*sample_args, **sample_kwargs):
+                val = trace(*clone_inputs((sample.input, *sample.args)), **sample.kwargs)
+                self.assertEqual(ref, val)
+
+                if op.expect_nnc_fusion:
```
if `expect_nnc_fusion` is False, do we want to skip the test i.e. tracing and comparing with eager? There must be other jit tests that test tracing?

davidberard98 replied:
@Krovatkin I think the reason we might want to run this on all ops (even the ones where nnc fusion isn't expected to occur) is that we still want to check that the output is the same after any nnc pass, regardless of whether or not we expect nnc to do anything

Krovatkin commented on test/test_ops.py:
```diff
@@ -1155,6 +1156,68 @@ def _fn(*sample_args, **sample_kwargs):
+                    continue
+
+                trace = create_traced_fn(self, variant)
+                ref = variant(*clone_inputs((sample.input, *sample.args)), **sample.kwargs)
```
can we also test `torch.jit.script`?

davidberard98 replied:
@Krovatkin I think the idea here was that more functions support tracing than scripting, so we're starting off with only testing with tracing

Krovatkin commented on test/test_ops.py:
```diff
@@ -1155,6 +1156,68 @@ def _fn(*sample_args, **sample_kwargs):
                 graph = traced.graph_for(*inp)
                 FileCheck().check(op_name).check_not(variant_name).run(graph)
 
+class TestJitTEFuser(JitTestCase):
```
I wonder if it would make sense to call this class something different so we can tell it apart from the one in `test_jit_fuser_te.py`?

Krovatkin reviewed (commented):
@davidberard98 to double check, we are adding another test class in addition to `TestTEFuser` in `test_jit_fuser_te.py`?

Krovatkin commented:
Adding a few folks from the NNC team, so we at least get one review from them. cc: @huiguoo @ZolotukhinM @navahgar

davidberard98 commented:
> @davidberard98 to double check, we are adding another test class in addition to `TestTEFuser` in `test_jit_fuser_te.py`?

@Krovatkin yes, for these reasons:
* ops_db contains a larger list of ops
* we're planning on doing this for NVFuser as well, so we figured it makes sense to add the same type of tests for NNC

ZolotukhinM commented:
Dont we already have OpInfo tests in NNC?
https://github.com/pytorch/pytorch/blob/bc026c05775a6504e48363941285ae381e52b931/test/test_jit_fuser_te.py#L2110-L2150

davidberard98 commented:
@ZolotukhinM oh cool, I didn't know about that test, thanks!

cc @eellison , is there anything else we would want here?

eellison commented:
The existing tests are only instantiated with a whitelisted set of operators, https://github.com/pytorch/pytorch/blob/bc026c05775a6504e48363941285ae381e52b931/test/test_jit_fuser_te.py#L2154, this would run more generally and get better coverage

davidberard98 commented:
@eellison we were going to do 2 things in these tests:

a) reuse op_db for an extensive list of ops to test nnc to make sure that input == output on nearly all ops (as long as they can be traced, etc.)
b) test on a smaller subset of op_db for which `expect_nnc_fusion == True` to make sure that nnc fusion occurs on these ops

Part B requires some extra work of making sure that expect_nnc_fusion is correctly set in all of the ops in op_db; it seems like this is already done in `test_jit_fuser_te.py`.

So in that case, what do you think about only testing for part A? We could extend test_jit_fuser_te.py tests if we wanted more coverage, and then we could avoid having to sync the list of nnc ops between `test_jit_fuser_te.py` and `op_db`

(also, regarding tests only running in float - not really sure what the expected behavior is, but this was going also going to be an issue for the test in this PR, since I saw fusion occurring in float didn't imply that it would occur in complex, and vice versa)

eellison commented:
Sure - starting with part A) makes sense to me. Although eventually it would be nice to merge the tests. Even just landing this then making a separate version of it that runs with nvfuser also would be nice

eellison reviewed (approved):
LGTM

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[WIP][JIT] Opinfo test for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
List of issues identified from these tests:
* addmm: JIT vs eager discrepancy: #71784 (note: may just be within expected numerical errors)
* uninitialized memory returned by some matmul ops, when inner dimensions are zero (fbgemm issue): #71774
* cat, batch_norm NNC implementations fail under some obscure conditions #71286
* NNC SimpleIREvaluatorImpl makes casts with undefined behavior ([here](https://github.com/pytorch/pytorch/pull/70465/files#diff-33783d984927670883fec7121b94a5142e54bedf159d7b85af6800818e513d09))
* Opinfo tests for mvlgamma sometimes provide invalid inputs #71794

ZolotukhinM commented:
> List of issues identified from these tests:
> 
> * addmm: JIT vs eager discrepancy: [[JIT] addmm differs from eager mode #71784](https://github.com/pytorch/pytorch/issues/71784) (note: may just be within expected numerical errors)
> * uninitialized memory returned by some matmul ops, when inner dimensions are zero (fbgemm issue): [matmul returns uninitialized memory for int64 tensors with inner dimension of zero #71774](https://github.com/pytorch/pytorch/issues/71774)
> * cat, batch_norm NNC implementations fail under some obscure conditions [[JIT][tensorexpr] cat, batch_norm opinfo tests failing #71286](https://github.com/pytorch/pytorch/issues/71286)
> * NNC SimpleIREvaluatorImpl makes casts with undefined behavior ([here](https://github.com/pytorch/pytorch/pull/70465/files#diff-33783d984927670883fec7121b94a5142e54bedf159d7b85af6800818e513d09))
> * Opinfo tests for mvlgamma sometimes provide invalid inputs [Opinfo test for mvlgamma: add epsilon #71794](https://github.com/pytorch/pytorch/pull/71794)

This is awesome! One comment and one question:

1) The observed behavior with `cat` is expected - `cat` only works if the list of the input tensors is created within the kernel too. We cannot codegen `cat` of a list of arbitrary tensors.

2) Do you have a repro for the SimpleIREval issue?

davidberard98 commented:
@ZolotukhinM 

If you have a build of pytorch with asan and -fsanitize=undefined, you can comment out the [__ubsan_ignore_undefined__](https://github.com/pytorch/pytorch/pull/70465/files#diff-33783d984927670883fec7121b94a5142e54bedf159d7b85af6800818e513d09) and then run `PYTORCH_TENSOREXPR_DONT_USE_LLVM=1 python3 test/test_jit_fuser_te.py -k test_nnc_correctness -v` and it should crash when it encounters failures. Unfortunately, I haven't been able to get ubsan running, so I've only found 3 failures via CI test failures, here #71849

If you can't get asan/ubsan running, you can check out: https://github.com/pytorch/pytorch/commit/51dde00e3f584e399d4be75db6143d50d7f0550c. In that commit I added asserts that check a subset of the ubsan checks.

And then run

```
PYTORCH_TENSOREXPR_DONT_USE_LLVM=1 python3 test/test_jit_fuser_te.py -k test_nnc_correctness_addcmul_cpu_int8 -v
```

See #71849 for other failures.

I also isolated one of the addcmul failures here (after looking more closely, it seems clear why this is failing - although ubsan doesn't find errors in eager mode for some reason):

```python
# repro_ubsan.py
# run PYTORCH_TENSOREXPR_DONT_USE_LLVM=1 python3 repro_ubsan.py
import torch

torch._C._jit_set_profiling_mode(True)
torch._C._jit_set_profiling_executor(True)
torch._C._jit_can_fuse_on_cpu()
torch._C._jit_set_texpr_fuser_enabled(True)
torch._C._debug_set_fusion_group_inlining(False)
torch._C._jit_set_te_must_use_llvm_cpu(False)

args = []
args.append(torch.tensor([[-4, -5, -9, -3,  3],
                          [ 0, -6, -5, -9, -8],
                          [ 3, -8,  4,  9,  7],
                          [ 8,  8, -5, -3,  3],
                          [-8,  8, -4, -5, -2]], dtype=torch.int8))
args.append(torch.tensor([[ 9, -9, -6,  6,  1],
                          [ 1, -7, -4,  0, -8],
                          [ 0, -8, -5,  5,  2],
                          [-1,  6, -1,  1, -4],
                          [-5, -2, -7,  5, -6]], dtype=torch.int8))
args.append(torch.tensor([[ 0, -4,  9, -7, -2],
                          [ 6,  6,  5, -4, -2],
                          [ 8,  1,  0,  7, -5],
                          [-3,  9,  6, -9, -7],
                          [ 0,  5,  1, -7, -4]], dtype=torch.int8))

a, b, c = args

print(torch.addcmul(a, b, c, value=3.14))

def f(a, b, c):
    return torch.addcmul(a, b, c, value=3.14)

tr = torch.jit.trace(f, (a, b, c))

tr(a, b, c)
tr(a, b, c)
```

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@eellison could you take one more look at the more recent changes?

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

eellison reviewed (approved):
great job !!

davidberard98 committed:
Update on "[JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

davidberard98 committed:
Update on "[JIT] Opinfo tests for nnc fusion"


These tests check to ensure that
(a) the result after nnc fusion (of a single op) is the same as the
unfused op
(b) for certain ops where fusion is expected to occur, ensure that
fusion does actually occur

Differential Revision: [D33595240](https://our.internmc.facebook.com/intern/diff/D33595240)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33595240).

malfet commented:
Reverting as it's caused OSS CI regressions which are consistent with the ones reported in PR CI, see https://github.com/pytorch/pytorch/runs/5070411422?check_suite_focus=true
